### PR TITLE
feat(linter): convert-tslint-to-eslint generators

### LINKED ISSUE #4320: Migration to ESlint
I have a nx workspace already configured with TSlint and I've migrated to nx 11 recently.
How do I migrate to ESLint? [angular-eslint](https://github.com/angular-eslint/angular-eslint#usage-with-nx-monorepos) instruct us to follow nx path in order to migrate properly. Which path would it be? I couldn't find it anywhere.

### Environment
``` Node : 12.14.0
  OS   : linux x64
  npm  : 6.13.4
  
  nx : Not Found
  @nrwl/angular : 11.0.4
  @nrwl/cli : 11.0.4
  @nrwl/cypress : 11.0.4
  @nrwl/devkit : 11.0.4
  @nrwl/eslint-plugin-nx : Not Found
  @nrwl/express : Not Found
  @nrwl/jest : 11.0.4
  @nrwl/linter : 11.0.4
  @nrwl/nest : Not Found
  @nrwl/next : Not Found
  @nrwl/node : Not Found
  @nrwl/react : Not Found
  @nrwl/schematics : Not Found
  @nrwl/tao : 11.0.4
  @nrwl/web : Not Found
  @nrwl/workspace : 11.0.4
  typescript : 4.0.5```

### PR DESCRIPTION
**Pull request by JamesHenry**
## Current Behavior

Nx supports creating Angular (and some others like NestJS) projects with both TSLint and ESLint as the linter, but with TSLint being EOL at the end of 2020, there is a need to be able to migrate existing TSLint projects over to ESLint.

This is not currently available in Nx.

## Expected Behavior

Nx provides a generator to assist with the conversion from TSLint to ESLint for Angular, Cypress and Nest projects.

## Related Issue(s)

Fixes #4320

### CODE REVIEW

**Comment by rmartin-sommset** [comment]
Thanks for all of your work on this @JamesHenry! 👍🏼

**Comment by pburkindine** [comment]
Awesome sauce!

**Comment by pburkindine** [comment]
This is huge!  Thanks!

**Comment by s4m0r4m4** [comment]
@FrozenPandaz can this be merged in? Looking to use this asap now that it's ready.

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]
Looking good :smiley:

File: packages/nest/src/generators/convert-tslint-to-eslint/schema.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        "$source": "argv",
+        "index": 0
+      },
+      "x-prompt": "Which NestJS project would you like to convert from TSLint to ESLint?"
+    },
+    "removeTSLintIfNoMoreTSLintTargets": {
+      "type": "boolean",
+      "description": "If this conversion leaves no more TSLint usage in the workspace, it will remove TSLint and related dependencies and configuration",
+      "default": true,
+      "x-prompt": "Would you like to remove TSLint and its related config if there are no TSLint projects remaining after this conversion?"
```
[the remark this answers is not in the document]
Comment (JamesHenry) [reply packages/nest/src/generators/convert-tslint-to-eslint/schema.json:28]: I'm not sure what kind of "shared state" you are referring to to achieve (2)? How would you change a default based on user behavior? Literally patch the schema in node_modules after the first run of the schematic?

File: packages/angular/index.ts [hunk 2]
```diff
@@ -6,3 +6,7 @@ export {
   pessimisticUpdate,
 } from './src/runtime/nx/data-persistence';
 export { NxModule } from './src/runtime/nx/nx.module';
+export {
```
[the remark this answers is not in the document]
Comment (JamesHenry) [reply packages/angular/index.ts:9]: Yep good catch!

File: packages/node/index.ts [hunk 3]
```diff
@@ -1,2 +1,3 @@
 export { applicationGenerator } from './src/schematics/application/application';
 export { libraryGenerator } from './src/schematics/library/library';
+export { addLintingToApplicationGenerator } from './src/schematics/add-linting-to-application/add-linting-to-application';
```
[the remark this answers is not in the document]
Comment (JamesHenry) [reply packages/node/index.ts:3]: Cool will remove it, thanks

File: packages/linter/src/utils/convert-tslint-to-eslint/project-converter.ts [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+     * and use the tslint-to-eslint-config package within the same execution.
+     *
+     * This is a worthwhile trade-off and the dry-run output doesn't offer a ton of value for this use-case anyway.
+     */
+    if (
+      process.argv.includes('--dry-run') ||
+      process.argv.includes('--dryRun') ||
+      process.argv.includes('-d')
+    ) {
+      throw new Error(
```
[the remark this answers is not in the document]
Comment (JamesHenry) [reply packages/linter/src/utils/convert-tslint-to-eslint/project-converter.ts:100]: Will do

File: packages/angular/src/generators/convert-tslint-to-eslint/convert-tslint-to-eslint.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+   */
+  projectConverter.removeProjectTSLintFile();
+
+  /**
+   * If the Angular project is an app which has an e2e project, convert that as well.
+   */
+  let cypressInstallTask: GeneratorCallback = () => Promise.resolve(undefined);
+  const e2eProjectName = projectConverter.getE2EProjectName();
+  if (e2eProjectName) {
+    cypressInstallTask = await cypressConversionGenerator(host, {
```
[the remark this answers is not in the document]
Comment (JamesHenry) [reply packages/angular/src/generators/convert-tslint-to-eslint/convert-tslint-to-eslint.ts:80]: Will do

File: packages/nest/src/generators/convert-tslint-to-eslint/schema.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (FrozenPandaz) [reply packages/nest/src/generators/convert-tslint-to-eslint/schema.json:28]: You can add the setting to `workspace.json` under `generators`.

```json
{
  ...,
  "generators": {
    "@nrwl/angular:convert-tslint-to-eslint": {
      "removeTSLintIfNoMoreTSLintTargets": options.removeTSLintIfNoMoreTSLintTargets
    }
  }
}
```

Here's an example: https://github.com/nrwl/nx/blob/00516ecd682fd3c58782bc498bdd91a380ed421f/packages/web/src/generators/application/application.ts#L163

Afterward, the setting will be used and the user won't be prompted.

[the remark this answers is not in the document]
Comment (JamesHenry) [reply packages/nest/src/generators/convert-tslint-to-eslint/schema.json:28]: Ahhh that does makes sense :)

**Comment by guibeg** [comment]
Hi @FrozenPandaz I wonder if there's an ETA for this pull request ? Thanks ! :)

**Comment by bopm** [comment]
@vsavkin I know that project has grown and requires a lot of hands and tons of time and I really appreciate all your combined work as a team on this project which saved me a lot of my time and nerves. But this way of interacting with end-users on a really serious problem (obviously given us by the Angular team's decision to postpone it almost to the point of no return but still) starting to revive a lot of memories of dealing with the Angular Core team over the years. Can something be done to improve that. How any of us can help with it?

**Comment by FrozenPandaz** [comment]
Everyone, thank you for being patient. We definitely understand that TSLint users need a migration path to ESLint.

@JamesHenry has done a wonderful job going through numerous iterations and a lot of exploration around how to accurately handle this very complicated conversion.

This is a priority and we are actively working on it. Rest assured, it _is_ coming. We have a few more things to iron out but I don't think it will be much longer.

As excited as you may be, please refrain from asking for concrete ETAs. Please continue being patient and give us some time and space to work out a _good_ solution. The last thing we want is to migrate users to a state which is hard to continue from.

**Review by FrozenPandaz — APPROVED** [review]

**Comment by jvandemo** [comment]
Great work, @JamesHenry!! 🎉 🏆

**Pull request merged**